submissions/dashakim/memory-game/script.js

```diff
@@ -0,0 +1,90 @@
+const section = document.querySelector("section");
+
+const imagesNames = function(currentImg) {
+  return currentImg
+    .split(".")
+    .slice(0, 1)
+    .join("")
+    .substr(4);
+};
+
+const createFrontFaceCard = function(div) {
+  const frontImage = document.createElement("img");
+  frontImage.classList.add("front-face");
+  frontImage.setAttribute("src", frontImageQuestion);
+  frontImage.setAttribute("alt", "icon-question");
+  div.appendChild(frontImage);
+};
+
+const createBackFaceCard = function(div, el) {
+  const backImage = document.createElement("img");
+  backImage.classList.add("back-face");
+  backImage.setAttribute("src", el);
+  backImage.setAttribute("alt", imagesNames(el));
+  div.appendChild(backImage);
+};
+
+const createCard = function(storage) {
+  storage.map(i => {
+    const divMemoryCard = document.createElement("div");
+    divMemoryCard.classList.add("memory-card");
+    section.appendChild(divMemoryCard);
+    divMemoryCard.dataset.framework = imagesNames(i);
+    createFrontFaceCard(divMemoryCard);
+    createBackFaceCard(divMemoryCard, i);
+  });
+};
+
+createCard(images);
+
+const card = document.querySelectorAll(".memory-card");
+let lockBoard = false;
+let hasFlippedCard = false;
+let firstCard, secondCard;
+
+function flipCard() {
+  if (lockBoard) return;
+  if (this === firstCard) return;
+  this.classList.add("flip");
+  if (!hasFlippedCard) {
+    hasFlippedCard = true;
+    firstCard = this;
+
+    return;
+  }
+
+  secondCard = this;
+  checkForMatch();
+}
+
+function checkForMatch() {
+  let isMatch = firstCard.dataset.framework === secondCard.dataset.framework;
+  isMatch ? disableCards() : unflipCards();
+}
+
+function disableCards() {
+  firstCard.removeEventListener("click", flipCard);
+  secondCard.removeEventListener("click", flipCard);
+  resetBoard();
+}
+function unflipCards() {
+  lockBoard = true;
+  setTimeout(() => {
+    firstCard.classList.remove("flip");
+    secondCard.classList.remove("flip");
+    resetBoard();
+  }, 500);
+}
+function resetBoard() {
+  [hasFlippedCard, lockBoard] = [false, false][(firstCard, secondCard)] = [
+    null,
+    null
+  ];
+}
+(function shuffle() {
+  card.forEach(card => {
+    let random = Math.floor(Math.random() * 16);
+    card.style.order = random;
+  });
+})();
+card.forEach(card => card.addEventListener("click", flipCard));
```

Comment:
Very important thing about event listeners is event delegation. Adding single listener to each element will significantly increase the number of event listeners on your page and their large number will increase the amount of memory consumed by your page. Which will lead to a decrease in productivity. Without delegating events, you would have to re-bind the event on each new element added to the page.
Try to add single event listener to memory cards parent.
Some stuff, that may help you: [Medium](https://medium.com/@stasonmars/%D0%B4%D0%B5%D0%BB%D0%B5%D0%B3%D0%B8%D1%80%D0%BE%D0%B2%D0%B0%D0%BD%D0%B8%D0%B5-%D1%81%D0%BE%D0%B1%D1%8B%D1%82%D0%B8%D0%B8%CC%86-%D0%B2-javascript-d91cbdd8916a), [MDN](https://developer.mozilla.org/en-US/docs/Web/API/Event/target)

Reply:
Yes, you absolutely right about parent's listeners, I used on task dom, but it was just one event element that I needed to change. Now, I have to select both elements and not clear understand how to do it. 
I tried something like this `section.addEventListener('click', flipCard)`
 but it doesn't work with the function "flipCard"
also something like this but it's also wrong   
```section.onclick = event => { let target = event.target; if (!target) return;  flipCard();}```

Reply:
Ok, you're definitely right, but if I put event on hole section, my `flip` class sets only on section. 
So, I rewrited just how I could)